src/utils/Encode.sol

```diff
@@ -0,0 +1,156 @@
+
+    /// Used to describe calls from a Gnosis Safe
+    enum Operation {
+        Call,
+        DelegateCall
+    }
+
+    /// Constants for Safe.execTransaction inputs we don't usee
+    uint256 constant SAFE_TX_GAS = 0;
+    uint256 constant BASE_GAS = 0;
+    uint256 constant GAS_PRICE = 0;
+    address constant GAS_TOKEN = address(uint160(0));
+    address payable constant REFUND_RECEIVER = payable(address(uint160(0)));
+
+    /// Dummy types and variables to facilitate syntax, e.g: `Encode.proxyAdmin.upgrade(...)`
+    enum EncProxyAdmin {
+        A
+    }
+    enum EncUpgradeableBeacon {
+        A
+    }
+    enum EncGnosisSafe {
+        A
+    }
+
+    EncProxyAdmin internal constant proxyAdmin = EncProxyAdmin.A;
+    EncUpgradeableBeacon internal constant upgradeableBeacon = EncUpgradeableBeacon.A;
+    EncGnosisSafe internal constant gnosisSafe = EncGnosisSafe.A;
+
+    /// @dev Creates a new, clean `MultisigCall[] storage` pointer, guaranteeing
+    /// any previous pointers will not be overwritten.
+    /// Since we're in a library, we have to use assembly+slot pointers, but the
+    /// high level version of this function equates to:
+    ///
+    /// uint _idx = storage.idx;
+    /// storage.idx++;
+    /// MultisigCall[] storage calls = storage.map[_idx];
+    /// return calls;
+    function newMultisigCalls() internal returns (MultisigCall[] storage) {
+        bytes32 _IDX_SLOT = IDX_SLOT;
+        bytes32 _MAP_SLOT = MAP_SLOT;
+
+        uint256 idx;
+        assembly {
+            idx := sload(_IDX_SLOT)
+            sstore(_IDX_SLOT, add(1, idx))
+        }
+
+        // fn pointer indirection fools the compiler into letting us have
+        // an uninitialized storage pointer
+        function() pure returns (mapping(uint => MultisigCall[]) storage) fn;
+        function() pure returns (uint) fn2 = func;
+        assembly {
+            fn := fn2
+        }
+        mapping(uint256 => MultisigCall[]) storage map = fn();
+        assembly {
+            map.slot := _MAP_SLOT
+        }
+
+        return map[idx];
+    }
+
+    function func() internal pure returns (uint256) {
+        return 0;
+    }
+
+    /// @dev Appends a call to a list of `MultisigCalls`, returning the original storage pointer
+    /// to facilitate call chaining syntax, e.g:
+    ///
+    /// calls
+    ///   .append(...)
+    ///   .append(...)
+    ///   .append(...);
+    function append(MultisigCall[] storage calls, address to, bytes memory data)
+        internal
+        returns (MultisigCall[] storage)
+    {
+        calls.push(MultisigCall({to: to, value: 0, data: data}));
+
+        return calls;
+    }
+
+    /// @dev Encodes a call to `ProxyAdmin.upgrade(proxy, impl)`
+    function upgrade(EncProxyAdmin, address proxy, address impl) internal pure returns (bytes memory) {
+        return abi.encodeCall(IProxyAdmin.upgrade, (ITransparentUpgradeableProxy(proxy), impl));
+    }
```

Comment:
IMO nit but it feels like it would be more straightforward to leave abi.encodeCall(IProxyAdmin, to, params)

but make a `zDeployedProxy` just cast to `ITransparentUpgradeableProxy` by default...

Reply:
im not super sure what you're asking for, but this is less about how `ZeusScript` uses it and more about how this works when writing upgrade scripts. for upgrade scripts this is much nicer.

i can share examples if you want